### ISSUE TITLE
Implement Paths and PathValue

### ISSUE THREAD

**Issue by Beraliv** [ts-essentials/ts-essentials#226]
The problems come from type-challenge [Typed Get](https://github.com/type-challenges/type-challenges/blob/master/questions/270-hard-typed-get/README.md)

Sometimes it's really convenient to extract the type from the another type without dealing with the path (it can be optional, consists of arrays, tuple or something else)

e.g. you have an array of elements, which has optional tuple, which has first element, and you need to extract that element
with `Get` type you would write it this way:

```ts
interface ComplexObject {
  elements: {
    labels: [{ title: string, description: string }];
  }[]
}

// { title: string, description: string }
Get<ComplexObject, 'elements.0.labels.0'>;
// { title: string, description: string }
Get<ComplexObject, 'elements[0]labels[0]'>;
```

I covered it in the topic – https://blog.beraliv.dev/2021-03-26-typed-get/

We can also have a look at the way `type-fest` made it working with square brackets – https://github.com/sindresorhus/type-fest/blob/main/source/get.d.ts

**Comment by Beraliv** [comment]
Hey @zhaoyao91! 👋

Sorry for writing you here 😅

Someone deleted my comment from here – https://github.com/sindresorhus/type-fest/issues/213

What are your requirements for `Path`?

I've seen the code that you sent:

```typescript
type Ensure<X, Y> = X extends Y ? X : never;
type SafeKey<T> = T extends string | number ? `${T}` : never;
type SafeValue<O, K> = K extends keyof O ? O[K] : never;

type Paths<O, P = ""> = O extends object
  ? {
      [K in keyof O]:
        | (P extends "" ? K : `${SafeKey<P>}.${SafeKey<K>}`)
        | Paths<O[K], P extends "" ? K : `${SafeKey<P>}.${SafeKey<K>}`>;
    }[keyof O]
  : never;

type Get<O, P extends Paths<O>> = P extends `${infer X}.${infer XS}`
  ? Get<SafeValue<O, X>, Ensure<XS, Paths<SafeValue<O, X>>>>
  : SafeValue<O, P>;
```

I plan to implement it here anyway, would be good to include your requirements there as well

**Comment by Beraliv** [comment]
Is planned here – https://github.com/ts-essentials/ts-essentials/pull/303

**Comment by Beraliv** [comment]
It will be named as `Paths` and `PathValue`

**Comment by Beraliv** [comment]
It's currently merged and will be released by the end of April as major

**Comment by Beraliv** [comment]
It's released just now as a part of [v10.0.0](https://github.com/ts-essentials/ts-essentials/releases/tag/v10.0.0)
